compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -1044,6 +1049,559 @@ fn build_fn_ptr_addr_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, self_ty: Ty<'t
     new_body(source, IndexVec::from_elem_n(start_block, 1), locals, sig.inputs().len(), span)
 }
 
+fn build_async_destructor_ctor_shim<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    self_ty: Ty<'tcx>,
+) -> Body<'tcx> {
+    // FIXME: implement cleanup branches
+
+    let span = tcx.def_span(def_id);
+    let Some(sig) = tcx.fn_sig(def_id).instantiate(tcx, &[self_ty.into()]).no_bound_vars() else {
+        span_bug!(span, "async_drop_in_place_raw with bound vars for `{self_ty}`");
+    };
+
+    let param_env = tcx.param_env_reveal_all_normalized(def_id);
+    let source_info = SourceInfo::outermost(span);
+
+    let mut locals = local_decls_for_sig(&sig, span);
+    let return_local = Local::new(0);
+    let self_ptr = Local::new(1);
+
+    #[derive(Clone, Copy)]
+    enum GlueStrategy<'tcx> {
+        /// Empty strategy should generate this:
+        ///
+        /// ```ignore (illustrative)
+        /// ready_unit()
+        /// ```
+        Empty,
```

Comment:
can we avoid this variant by never requesting the shim for these at all?

Reply:
We would still need to be able to directly call `async_drop_in_place(_raw)::<()>` so we cannot "never request". This is true for regular drop glue too.

Reply:
May be relevant: https://github.com/rust-lang/rust/pull/121801#discussion_r1570638718